kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ProduceRequest.java

```diff
@@ -0,0 +1,305 @@
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.BinaryNode;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+import io.confluent.kafka.serializers.subject.RecordNameStrategy;
+import io.confluent.kafka.serializers.subject.TopicNameStrategy;
+import io.confluent.kafka.serializers.subject.TopicRecordNameStrategy;
+import io.confluent.kafka.serializers.subject.strategy.SubjectNameStrategy;
+import io.confluent.kafkarest.entities.EmbeddedFormat;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+@AutoValue
+public abstract class ProduceRequest {
+
+  ProduceRequest() {
+  }
+
+  @JsonProperty("partition_id")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<Integer> getPartitionId();
+
+  @JsonProperty("headers")
+  @JsonInclude(Include.NON_EMPTY)
+  public abstract ImmutableList<ProduceRequestHeader> getHeaders();
+
+  @JsonProperty("key")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<ProduceRequestData> getKey();
+
+  @JsonProperty("value")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<ProduceRequestData> getValue();
+
+  @JsonProperty("timestamp")
+  @JsonInclude(Include.NON_ABSENT)
+  public abstract Optional<Instant> getTimestamp();
+
+  @JsonCreator
+  static ProduceRequest fromJson(
+      @JsonProperty("partition_id") @Nullable Integer partitionId,
+      @JsonProperty("headers") @Nullable List<ProduceRequestHeader> headers,
+      @JsonProperty("key") @Nullable ProduceRequestData key,
+      @JsonProperty("value") @Nullable ProduceRequestData value,
+      @JsonProperty("timestamp") @Nullable Instant timestamp) {
+    return builder()
+        .setPartitionId(partitionId)
+        .setHeaders(headers != null ? headers : ImmutableList.of())
+        .setKey(key)
+        .setValue(value)
+        .setTimestamp(timestamp)
+        .build();
+  }
+
+  public static Builder builder() {
+    return new AutoValue_ProduceRequest.Builder().setHeaders(emptyList());
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    public abstract Builder setPartitionId(@Nullable Integer partitionId);
+
+    public abstract Builder setHeaders(List<ProduceRequestHeader> headers);
+
+    public abstract Builder setKey(@Nullable ProduceRequestData key);
+
+    public abstract Builder setValue(@Nullable ProduceRequestData value);
+
+    public abstract Builder setTimestamp(@Nullable Instant timestamp);
+
+    public abstract ProduceRequest build();
+  }
+
+  @AutoValue
+  public abstract static class ProduceRequestHeader {
+
+    ProduceRequestHeader() {
+    }
+
+    @JsonProperty("name")
+    public abstract String getName();
+
+    @JsonIgnore
+    public abstract Optional<ByteString> getValue();
```

Comment:
using a protobuf type here seems a bit out of place to me (i could be missing a good reason for it)

Reply:
`ByteString` is one of the best `immutable byte-array` types out there. (They can't make it part of Guava for legacy reasons). byte-array is not a class to start with, so it does not implement equals/hash-code properly. It also lacks reasonable constructors, like `copyFromUt8`. And, most importantly, it's mutable, which I limit as much as possible everywhere in Kafka REST.

Reply:
ok, those reasons sound sufficient ..